tutorials/3d/3d_rendering_limitations.rst

```diff
@@ -0,0 +1,125 @@
+.. _doc_3d_rendering_limitations:
+
+3D rendering limitations
+========================
+
+Introduction
+------------
+
+Due to their focus on performance, real-time rendering engines have many
+limitations. Godot renderer's is no exception. To work effectively with those
+limitations, you need to understand them.
+
+Texture size limits
+-------------------
+
+Modern desktop/laptop GPUs can only render textures up to 16384×16384. If you
+need to display a larger texture for some reason, you'll need to split it into
+multiple smaller textures. Old GPUs like the GeForce 8/9 series are limited to
+8192×8192.
+
+Mobile GPUs are typically limited to 4096×4096 textures. Therefore, if you want
+your texture to display correctly on all platforms, you should avoid using
+textures larger than 4096×4096.
+
+2D line drawing
+---------------
+
+In OpenGL, 2D line drawing is implemented by the graphics driver. The maximum
+line thickness that can be drawn is implementation-defined. AMD and Intel
+graphics drivers only support 1 pixel-thick lines. NVIDIA graphics drivers allow
+drawing lines up to 5 pixels thick.
+
+Note that this limitation doesn't affect lines drawn by the :ref:`class_Line2D`
+node, since these are drawn using triangle strips instead of relying on OpenGL
+line drawing.
+
```

Comment:
If this is about 2D line drawing, and `Line2D` isn't affected, where *does* this affect me as a Godot user?

Reply:
This is mainly relevant for CanvasItem's `draw_line()` methods (and `draw_multiline()`, `draw_polyline()`, ...).

Reply:
I suggest adding a note about that then, because I wasn't sure where it applied. :)

Reply:
Actually, I'm not even sure if this is relevant in Godot's context. Godot will automatically use rectangles instead of lines for lines thicker than 1 pixel.

cc @lawnjelly, as he worked on https://github.com/godotengine/godot/pull/38417.

Reply:
As far as I can see for 2D TYPE_LINE is drawn with GL_LINES for size 1 and below, above that it uses a combination of rects and an optional 4 GL_LINES around the edges for anti-aliasing.

The option `draw_lines_as_rects` that I added just always uses rects and removes any anti aliasing. So it should run faster on hardware that has poor GL_LINES support, at the cost of not looking as good.

TYPE_POLYLINE I'm not that familiar with and it is hardly used in the editor, but it seems to use a bunch of polys followed by a GL_LINE_LOOP, or a bunch of GL_LINES, depending on whether there are triangles.

It's hard to give that much advice on lines, except to say anything using GL_LINES has the potential to be very slow on some hardware. It is regarded as 'ye olde' OpenGL and some have given up on supporting it. On the whole I've heard it recommended to only use GL_LINES for debugging. For drawing UI lines, and polys with lines around in modern games I get the impression using polys and shaders is a more favoured approach.

It is possible in the future we can modify the TYPE_POLYLINE to do this a bit more sensibly with `draw_lines_as_rects` set. And have a look at line shaders.

So really the whole lines area is subject to change, it might become quite safe to use in the future, but it's could be a little bit risky at the moment. So far the only hardware we've been hearing about with possible slow lines is Mac, but that is still waiting for confirmation.

Also, it may be slow on e.g. phones but we don't know as few people write games with lines.

Reply:
I removed the 2D line drawing section since it's not relevant in Godot's context.